src/ctelnet.cpp

```diff
@@ -1413,6 +1441,91 @@ void cTelnet::processTelnetCommand(const std::string& command)
+
+            // payload is in the Mud Server's encoding, trim off the Telnet suboption
+            // bytes from beginning (3) and end (2):
+            payload = payload.mid(3, static_cast<int>(payload.size()) - 5);
+
+            QString transcodedMsg;
+
+            if (mpOutOfBandDataIncomingCodec) {
+                // Message is encoded
+                transcodedMsg = mpOutOfBandDataIncomingCodec->toUnicode(payload);
+            } else {
+                // Message is in ASCII (though this can handle Utf-8):
+                transcodedMsg = QString::fromUtf8(payload);
+            }
+
+            // remove \r and \n from the data.
+            transcodedMsg.remove(QChar::CarriageReturn);
+            transcodedMsg.remove(QChar::LineFeed);
+            // replace ANSI escape character with escaped version, to handle improperly passed ANSI codes
+            transcodedMsg.replace(QLatin1String("\u001B"), QLatin1String("\\u001B"));
+
+            // CHARSET support per https://tools.ietf.org/html/rfc2066
+            if (command[3] == CHARSET_REQUEST) {
+                // No translate table support.  Discard. 
+                if (transcodedMsg.startsWith(QStringLiteral("[TTABLE]1"))) {
+                    transcodedMsg.remove(QStringLiteral("[TTABLE]1"));
+                }
+
+                // Second character is the separator.
+                QChar separator = transcodedMsg[1];
+                QStringList characterSetList = transcodedMsg.split(separator);
+                QString acceptedCharacterSet;
+
+                if (characterSetList.size() > 0) {
+                    for (int i = 1; i < characterSetList.size(); ++i) {
+                        if (mAcceptableEncodings.contains(characterSetList[i].toLatin1()) || mAcceptableEncodings.contains(("M_" + characterSetList[i]).toLatin1())) {
+                            acceptedCharacterSet = characterSetList[i];
+                            break;
+                        }
+                    }
+                }
+
+                std::string output;
+                output += TN_IAC;
+                output += TN_SB;
+                output += OPT_CHARSET;
+
+                if (!acceptedCharacterSet.isEmpty()) {
+                    setEncoding(acceptedCharacterSet.toLatin1(), true);
+
+                    output += CHARSET_ACCEPTED;
+                    output += encodeAndCookBytes(acceptedCharacterSet.toStdString());
+                } else {
+                    output += CHARSET_REJECTED;
+                }
+
+                output += TN_IAC;
+                output += TN_SE;
+                socketOutRaw(output);
+            } else if (command[3] == CHARSET_ACCEPTED) {
+                // Case unlikely.  Mudlet does not initiate negotiations yet.  Do nothing.
+            } else if (command[3] == CHARSET_REJECTED) {
+                // Case unlikely.  Mudlet does not initiate negotiations yet.  Do nothing.
+            } else if (command[3] == CHARSET_TTABLE_IS) {
```

Comment:
Ah, perhaps this will be enough to reject the `[TABLE]` bit I worried about further up...